### DISCUSSION TITLE
Incorrect conversion on Union of Iterables?

### DISCUSSION THREAD

**Discussion by ringohoffman** [microsoft/pyright#5198] [Q&A] [answered]
In the following case:

```python
from typing import Iterable, Union

def f(x: Union[Iterable[int], Iterable[str]]) -> None:
    x = list(x)  # Expression of type "list[int | str]" cannot be assigned to declared type "Iterable[int] | Iterable[str]"

```

wouldn't the expression `list(x)` have the type `list[int] | list[str]`, not `list[int | str]`?

Pylance `v2023.5.40` for VS Code.

**Comment by erictraut** [comment, marked as answer]
Type checkers typically iterate through the subtypes of a union at the point at which they're used. In this case, pyright iterates through the `Iterable[int]` and `Iterable[str]` at the point where it's evaluating the expression `x` passed to the `list` constructor. It doesn't evaluate the broader expression `list(x)` multiple times because one of its arguments happens to be a union. Doing so wouldn't be feasible because expressions can be deeply nested.

Not surprisingly, mypy generates the same error as pyright here.

So, this is working as designed. I don't consider this a bug.
